rslib/src/cloze.rs

```diff
@@ -31,99 +19,262 @@ lazy_static! {
+    for c in text.chars() {
+        match c {
+            '"' => out.push_str("&quot;"),
+            '&' => out.push_str("&amp;"),
+            '\n' => out.push_str("&NewLine;"),
+            '<' => out.push_str("&lt;"),
+            '>' => out.push_str("&gt;"),
+            _ => out.push(c),
+        }
+    }
+
+    Cow::from(out)
+}
+
+/// Parse string for clozes and return:
+/// cloze_only == false: resulting HTML string
+/// cloze_only == true: resulting HTML for the cloze texts only
+/// no cloze found: empty string
+fn reveal_clozes(text: &str, cloze_ord: u16, question: bool, cloze_only: bool) -> Cow<str> {
+    let mut state = State::Root;
+    let mut stack: Vec<Cloze> = vec![Cloze::new()];
+    let mut current_found = false;
 
-        let text_attr;
-        let replacement;
-        if question {
-            text_attr = format!(r#" data-cloze="{}""#, htmlescape::encode_attribute(&text));
-            // hint provided?
-            if let Some(hint) = caps.get(cloze_caps::HINT) {
-                replacement = format!("[{}]", hint.as_str());
-            } else {
-                replacement = "[...]".to_string();
+    for c in text.chars() {
+        let old_state = state;
+        state = process_char(state, c);
+
+        match state {
+            State::Open => stack.push(Cloze::new_from(c)),
+            State::Close => {
+                current_found |= close(&mut state, &mut stack, cloze_only, cloze_ord, question);
+            }
+            State::Abandon => abandon(&mut state, &mut stack, c),
+            _ => {
+                let last = stack.last_mut().unwrap();
+                last.buf.push(c);
+                match state {
+                    State::Root => {
+                        if !cloze_only {
+                            last.text.push(c);
+                        }
+                    }
+                    State::Open2 => {
+                        if old_state == State::Open2 {
+                            shift_open(&mut stack)
+                        }
+                    }
+                    State::Ord => last.ord_str.push(c),
+                    State::Text => last.text.push(c),
+                    State::Hint => last.hint.push(c),
+                    _ => {}
+                }
             }
-        } else {
-            text_attr = "".to_string();
-            replacement = text;
         }
+    }
 
-        format!(r#"<span class="cloze"{}>{}</span>"#, text_attr, replacement)
-    });
+    if !current_found {
+        return Cow::Borrowed("");
+    }
+    if cloze_only {
+        return stack.last().unwrap().text[2..].to_owned().into();
+    }
 
-    if !cloze_ord_was_in_text {
-        return "".into();
+    if stack.len() > 1 {
+        let cloze = stack.pop().unwrap();
+        stack.last_mut().unwrap().text.push_str(&cloze.buf);
+    }
+    return Cow::from(stack.last().unwrap().text.to_owned());
+
+    // 3 consecutive {{{, shift the first onto the "parent" text
```

Comment:
Why?

Reply:
Not sure I understand which part you are wondering about? If it's the `stack.last_mut().unwrap().text.push_str(&cloze.buf);` it is to keep any remaining "open" cloze (i.e. `foo {{c1::bar`) as "text", although looking at it now that should probably be a while loop to handle `{{c1::foo{{c2::bar{{c3:baz`

Reply:
Sorry, I was referring to the comment.